### PR TITLE
WIP - RECAP: Implementation from Physical Intelligence Pistar06

### PR DESCRIPTION
**Pull request by s1lent4gnt**
## Title

Implementation of RECAP Physical Intelligence Pistar06 paper

Paper: https://www.pi.website/download/pistar06.pdf
Blog: https://pi.website/blog/pistar06

## Summary / Motivation

This PR starts the full RECAP implementation in **LeRobot**, beginning with the distributional value function and continuing through the TODO list below.

## Related issues

- Fixes / Closes: N/A
- Related: #3076

## What changed

- Added distributional value function as a reward model:
  - `configuration_distributional_value_function.py`
  - `modeling_distributional_value_function.py`
  - `processor_distributional_value_function.py`
  - `__init__.py` exports + factory wiring
  - Optional `recap` dependency group in `pyproject.toml`
  - Unit tests `tests/rewards/test_distributional_value_function.py`

## Implementation details
This implementation follows the pi*0.6 value-function recipe, with explicit engineering choices documented below.
- **Backbone choice**
  - Paper: language-conditioned distributional value function with a smaller (~670M) VLM backbone, initialized from Gemma 3 VLM weights.
  - Here: PI05-compatible PaliGemma stack (6 LM + 13 vision layers).
- **Initialization path**
  - Paper: value function initialized from a pre-trained VLM and iteratively fine-tuned.
  - Here: initialize from PI05 actor checkpoint on first run (`init_from_actor_path`), then continue from saved value-function checkpoints.
- **Distributional training target**
  - Paper: discretize empirical returns into `B=201` bins and train with cross-entropy (Eq. 1).
  - Here: default target uses HL-Gauss smoothing.  
  Farebrother et al., *"Stop Regressing: Training Value Functions via Classification for Scalable Deep RL"* (2024)  
  https://arxiv.org/abs/2403.03950,
**optional** `dirac_delta` projection Distributional RL / categorical value bins (C51 reference)
Bellemare et al., "A Distributional Perspective on Reinforcement Learning" (2017)
https://arxiv.org/abs/1707.06887, and
**optional** one-hot terminal override are provided as configurable alternatives.
- **Data contract**
  - Here: `mc_return` and `is_terminal` are consumed as dataset columns in the flat training batch dict and are produced by the annotation pipeline (`returns` module), not inferred inside the value model.
- **Reward/return shaping**
  - Paper: sparse reward design (`-1/step`, `0/success`, `-C_fail/failure`) and per-task normalization of values to approximately `(-1, 0)`.
  - Here: same reward semantics are implemented via dataset annotation and used to compute Monte Carlo returns for value-function training.
### Community credit
- Big thanks to **@cijerezg** for his open-source implementation [here](https://github.com/cijerezg/lerobot)

## TODO (RECAP end-to-end)
- [ ] **Distributional Value Function**
  - [x] Config + registry + factory wiring
  - [x] Model architecture (6 LM, 13 vision, 201 bins)
  - [x] Init path finalized (first run from PI05 actor; subsequent runs from VF checkpoint)
  - [x] Target modes (HL-Gauss default, Dirac delta optional, terminal one-hot optional)
  - [x] Batch contract (`mc_return`, `is_terminal` from dataset columns in flat batch)
  - [x] Processor (task prompt, tokenizer, image preprocess)
  - [ ] Tests
- [ ] **Returns Annotation (`lerobot-annotate --modules=returns`)**
  - [ ] Sparse reward computation (`-1/step`, `0/success`, `-C_fail/failure`)
  - [ ] Episode-aware `is_terminal` generation
  - [ ] MC return computation + normalization by task max episode length
  - [ ] Dataset column writing + schema validation
- [ ] **Advantage Annotation (`lerobot-annotate --modules=advantage`)**
  - [ ] Advantage computation implemented for required modes (pretrain/fine-tune settings)
  - [ ] Binarization (`I_t`) and threshold policy implemented
  - [ ] Annotation write path integrated with language pipeline (`style="advantage"`)
  - [ ] Recipe YAML composition validated with subtasks/task_aug/advantage
- [ ] **Policy Conditioning (Conditional SFT)**
  - [ ] Prompt injection (`Advantage: positive/negative`) added before tokenization
  - [ ] 30% conditioning dropout
- [ ] **CFG Inference (Optional)**
  - [ ] Eq. 13 interpolation path: `v_uncond + β * (v_cond - v_uncond)`
  - [ ] Hidden-state caching wired (VLM cached; action expert run twice per denoising step)
  - [ ] Beta controls exposed (default `β=1`, optional `β>1` settings)
- [ ] **Data Collection (DAgger + labels)**
  - [ ] `lerobot-rollout --strategy.type=dagger` integration validated for RECAP loop
  - [ ] Episode success labeling persisted for returns module
  - [ ] Intervention tagging compatibility validated with downstream annotation steps
- [ ] **Iterative RECAP Loop**
  - [ ] End-to-end loop : train V → score advantages → train policy → collect data → repeat
- [ ] **Validation and Documentation**
  - [ ] Unit/integration test coverage finalized across components
  - [ ] PR/docs updated with commands and expected artifacts

## How was this tested (or how to run locally)

- Tests added
```tests/rewards/test_distributional_value_function.py```
- Local checks:
```bash
uv run pytest tests/rewards/test_distributional_value_function.py -q
```

## Checklist (required before merge)

- [ ] Linting/formatting run (`pre-commit run -a`)
- [ ] All tests pass locally (`pytest`)
- [ ] Documentation updated
- [ ] CI is green

## Reviewer notes

- Anything the reviewer should focus on (performance, edge-cases, specific files) or general notes.
- Anyone in the community is free to review the PR.

### CODE REVIEW

**Comment by paulpacaud** [comment]
Hello, this PR looks very promising, would you have an ETA of when it will be ready to test?
